[WIP] refactor: change lang selector logic

dcoa opened the pull request:
### Description and motivation

This PR refactors the lang selector logic recognizing its significance for multi-language instances.
The Open edX platform previously featured a configurable language selector in both the header and footer. However, after transitioning to microfrontends, this feature is only partially implemented in the footer. It requires passing a list of supported languages and a change function, necessitating additional modifications even when using the footer slot.

This PR change the dropdown and add the user preference update endpoint.

>[!WARNING]
> This depends on https://github.com/openedx/frontend-platform/pull/786

### Proposal 
Convert this component in a slot so could be installed in the header as well and give them the operators the option to use the language selector in the header or footer.  

### How to test

1. Override the Footer component in a MFE or use the example application to test the current implementation. 
3. You need the messages for the selected language if you want to visualize the change.
4. You can see the change in the language cookie. 


https://github.com/user-attachments/assets/5e224cde-9417-422e-baf9-3d6318ffc541

openedx-webhooks commented:
Thanks for the pull request, @dcoa!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.


#### :radio_button: Update the status of your PR

Your PR is currently marked as a draft.  After completing the steps above, update its status by clicking "Ready for Review", or removing "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->


#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?



This repository is currently unmaintained.

To get help with finding a technical reviewer, tag the community contributions project manager for this PR in a comment and let them know that your changes are ready for review:

1. On the right-hand side of the PR, find the Contributions project, click the caret in the top right corner to expand it, and check the "Primary PM" field for the name of your PM.
2. Find their GitHub handle [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3548807177/Community+Contributions+Project+Manager#Current-OSPR-Project-Managers).


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

dcoa committed:
refactor: change lang selector logic

dcoa committed:
refactor: separate the test for lang seector

dcoa commented:
Hi, @brian-smith-tcril I has wondering if you can have a look to this proposal or if should I tag someone from product before the technical review?

brian-smith-tcril commented:
Sorry for the extended delay in response!

I think this is a change we do want. I also know that there is a similar feature implemented in `frontend-base` at the moment. Ideally the implementation of this feature will align with what is needed by the feature in `frontend-base`. I think @arbrandes has a better sense of what the `frontend-base` implementation is at the moment, and what we'll want it to be by the time we are ready to start using `frontend-base`.

Assuming this PR and the implementation in `frontend-base` play well together I'd be more than happy to review it!

sarina commented:
> should I tag someone from product before the technical review

I think there's a question of whether or not we want to be adding a new flag. If we don't mind a new flag this PR is good to go from a product perspective, but if we want to make this the default experience for any site that has multiple languages, you'd want to make a product ticket.

dcoa commented:
@arbrandes I think your opinion here is valuable, understanding you have a better panorama of the current state of the frontend new architecture.

What do you think about this proposal and the alignment with `frontend-base`   also the introduction of new config variables?

mphilbrick211 commented:
> @arbrandes I think your opinion here is valuable, understanding you have a better panorama of the current state of the frontend new architecture.
> 
> What do you think about this proposal and the alignment with `frontend-base` also the introduction of new config variables?

Friendly ping on this, @arbrandes

dcoa commented:
@arbrandes any opinion in this topic?

Ang-m4 commented:
Hi @dcoa,
I hope you're doing well!
I came across this PR while working with the **NELC** team, as we were looking for a simple and practical implementation of a language selector. We really appreciate your proposal and are excited about the possibility of using it. We're also looking forward to seeing it merged upstream.
Additionally, we're currently working on a header implementation that follows the same logic. I’d appreciate it if you could take a look at it soon.

I’ll be leaving some comments on your code for further discussion, and i really would like to @arbrandes to take a look at this as he has a better understanding of the entire project.

Thank you for your hard work and contributions!

Ang-m4 commented on src/components/LanguageSelector/index.jsx:
```diff
@@ -0,0 +1,82 @@
+};
+const getLocaleName = (locale) => {
+  const langName = new Intl.DisplayNames([locale], { type: 'language', languageDisplay: 'standard' }).of(locale);
+  return langName.replace(/^\w/, (c) => c.toUpperCase());
```
```suggestion
  return langName.charAt(0).toUpperCase() + langName.slice(1);
```
I see that this line capitalizes the string. A more straightforward implementation without regular expressions might be better for this task. I'll leave it up to you.

Ang-m4 commented on src/components/LanguageSelector/index.test.jsx:
```diff
@@ -0,0 +1,141 @@
+    });
+  });
+
+  it('should disp,ay the language icon and modify the label according to the screen size', () => {
```
```suggestion
  it('should display the language icon and modify the label according to the screen size', () => {
```

Ang-m4 commented on src/components/LanguageSelector/data.js:
I would like you to take a look at this [PR](https://github.com/openedx/frontend-platform/pull/786), you might find it interesting as it proposes a new `frontend-platform` module to handle the language change logic based on your changes.

dcoa replied:
Hi @arbrandes , I would like to take this conversation back to the table, there is this proposal of centralize the logic for update the language in `frontend-platofrm` https://github.com/openedx/frontend-platform/pull/786 I think is a good idea, have the core logic agnostic to a component, specially if we want to extended the functionality to the header as well (as we used to have in comprehensive theming). I look forward to hearing your opinion.

Ang-m4 commented:
Hi @dcoa,

I've tested the final version of the component using `frontend-app-learning` and it looks great visually! However, I noticed an issue: when a new language is selected from the dropdown, it only updates the cookie in LMS, which prevents the application from loading and applying the correct translations. This happens because the `getLocale()` method (also used by `handleRtl()`) relies on the `LANGUAGE_PREFERENCE` cookie to retrieve its information. I recommend moving the cookie assignment outside of the conditional statement.

Please see the attached video demonstrating the behavior:  

https://github.com/user-attachments/assets/1822b7ba-e18c-4879-bb4b-b23ddd362dc4

After the change, everything seems to work fine:  

https://github.com/user-attachments/assets/54a5fbfc-20ae-4fed-b6a0-a6b1ed0364bb

However, please note that while the translations do update, they are only applied partially. Many sections of the platform are still not configured to receive the `publish()` event or handle translations dynamically.

Given this, the most robust solution might be to force a full page reload when updating the language, ensuring that all parts of the application reflect the new language settings.

What do you think?

Ang-m4 commented on src/components/LanguageSelector/index.jsx:
```diff
@@ -0,0 +1,82 @@
+import { patchPreferences, postSetLang } from './data';
+
+const onLanguageSelected = async ({ langCookieName, username, selectedLlocale }) => {
+  try {
+    if (username) {
+      await patchPreferences(username, { prefLang: selectedLlocale });
+      await postSetLang(selectedLlocale);
+    } else {
+      getCookies().set(langCookieName, selectedLlocale);
+    }
```
```suggestion
  try {
    getCookies().set(langCookieName, selectedLlocale);
    if (username) {
      await patchPreferences(username, { prefLang: selectedLlocale });
      await postSetLang(selectedLlocale);
    }
```

dcoa replied:
I thing the real problem is there are 2 cookies one is being updated the second one not, I will check if I can replicate the behavior and understand why that is happening.

dcoa replied:
The suggested change creates a new cookie instead of changing the default one, because of that is not the best solution for the described problem. 

I could identify that the described problem is only present in learning (that make sense for the iframe - django based does not change at runtime as react does).

arbrandes commented on src/components/Footer.test.jsx:
```diff
@@ -27,27 +27,25 @@ const FooterWithContext = ({ locale = 'es' }) => {
+    authenticatedUser,
     config: {
+      ENABLE_FOOTER_LANG_SELECTOR: true,
+      LANGUAGE_PREFERENCE_COOKIE_NAME,
```
This should be implemented as part of https://github.com/openedx/frontend-platform/pull/786.  This PR should only be concerned with the UI elements.

arbrandes reviewed (changes requested):
This is a good direction, but I'd like to see this refactored to take into account https://github.com/openedx/frontend-platform/pull/786.  As a matter of fact, both PRs should be developed in tandem, so that all the state management and API calls are handled by frontend-platform, and the footer need only concern itself with the presentation.

dcoa committed:
refactor: make the lang selector a slot

fix: update test and linter

docs: add slot readme example

test: update snapshoots

vv

dcoa commented:
I will pass this to draft and reopen it once the refactor is done

dcoa committed:
docs: add modify prop example

mphilbrick211 commented:
Hi @dcoa - just checking to see if you're planning to pursue this pull request? If not, we can close for now.

dcoa commented:
Understanding that we decided to move this forward in `frontend-base` ecosystem first  (see https://github.com/openedx/public-engineering/issues/558) I am going to close this PR and reopen once we have all defined and we can just backport the changes.
